src/peer.rs

```diff
@@ -57,149 +57,187 @@ pub(crate) async fn server_peer_loop(peer_id: PeerId, broker_writer: Sender<Brok
-            let mut result = shvproto::Map::new();
-            result.insert("nonce".into(), RpcValue::from(&nonce));
-            result.insert("azureClientId".into(), azure_config.client_id.as_str().into());
-            frame_writer.send_result(resp_meta, result.into()).await?;
-            nonce
-        };
-        {
-            let frame = frame_reader.receive_frame().await?;
-            let rpcmsg = frame.to_rpcmesage()?;
-            let resp_meta = RpcFrame::prepare_response_meta(&frame.meta)?;
-            if rpcmsg.method().unwrap_or("") != "login" {
-                frame_writer.send_error(resp_meta, "Invalid login message.").await?;
-                continue;
-            }
-            debug!("Client ID: {peer_id}, login received.");
-            let params = rpcmsg.param().ok_or("No login params")?.as_map();
-            let login = params.get("login").ok_or("Invalid login params")?.as_map();
-            user = login.get("user").ok_or("User login param is missing")?.as_str().to_string();
-            let password = login.get("password").ok_or("Password login param is missing")?.as_str();
-            let login_type = login.get("type").map(|v| v.as_str()).unwrap_or("");
+    let mut nonce = None;
+    'login_loop: loop {
+        let frame = frame_reader.receive_frame().await?;
+        let rpcmsg = frame.to_rpcmesage()?;
+        let resp_meta = RpcFrame::prepare_response_meta(&frame.meta)?;
+        let method = rpcmsg.method().unwrap_or("");
+        match method {
+            "hello" => {
+                debug!("Client ID: {peer_id}, hello received.");
+                let nonce: &String = nonce.get_or_insert_with(|| Alphanumeric.sample_string(&mut rand::rng(), 16));
+                let mut result = shvproto::Map::new();
+                result.insert("nonce".into(), RpcValue::from(nonce));
+                frame_writer.send_result(resp_meta, result.into()).await?;
+            },
+            "workflows" => {
+                debug!("Client ID: {peer_id}, workflows received.");
+                let mut workflows = make_list!{
+                    "PLAIN",
+                    "SHA1",
+                };
+                if let Some(azure_config) = &azure_config {
+                    workflows.push(make_map!{
+                        "type" => "oauth2-azure",
+                        "clientId" => azure_config.client_id.clone(),
+                        "authorizeUrl" => azure_config.authorize_url.clone(),
+                        "tokenUrl" => azure_config.token_url.clone(),
+                        "scopes" => azure_config.scopes.clone(),
+                    }.into());
+                };
+                frame_writer.send_result(resp_meta, workflows.into()).await?;
+            },
+            "login" => {
+                debug!("Client ID: {peer_id}, login received.");
+                let params = rpcmsg.param().ok_or("No login params")?.as_map();
```

Comment:
I think that an error should be sent back to the peer instead of just closing the connection.

Reply:
Moved to a new issue: #29